doc/carrierwave.md

```diff
@@ -326,6 +327,20 @@ instead of CarrierWave, using equivalent Shrine storages. For help with
 translating the code from CarrierWave to Shrine, you can consult the reference
 below.
 
+You'll notice that if you add any metadata manually or using plugins within
+your Shrine uploader, the metadata will not be filled in after using the above
+migration script. You can either modify the synchroniziation class above or use
+the `refresh_metadata` plugin and run `.refresh_metadata!` on all uploaded
+files:
+
+```rb
+Photo.find_each do |photo|
+  Shrine.uploaded_file(photo.image) do |uploaded_file|
+    uploaded_file.refresh_metadata!
```

Comment:
I actually cannot get this to work. I followed [the documentation](https://github.com/janko-m/shrine/blob/247f75ffd2cc27051c74337dda2da17c1027d35c/lib/shrine/plugins/versions.rb#L54-L62) but seem to be doing something wrong.

```rb
Shrine.uploaded_file(message_image.image) { |ul| puts ul.size }
Shrine::Error: {:thumb=>#<ImageUploader::UploadedFile:0x00007fa8b2675ae0 @data={"id"=>"private/uploads/message_image/image/12150/thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "storage"=>"store", "metadata"=>{"size"=>25772, "width"=>276, "height"=>368, "sha256"=>"ccfe5199e267fd45540d09dac6b85c5893978ea314a5493a2a86a36825875eb5", "filename"=>"thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "mime_type"=>"image/jpeg"}}>, :original=>#<ImageUploader::UploadedFile:0x00007fa8b2675978 @data={"id"=>"private/uploads/message_image/image/12150/b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "storage"=>"store", "metadata"=>{"size"=>550447, "width"=>1440, "height"=>1920, "sha256"=>"04b3c07231f78134998b624a3f6e00d64fb9b1755f60f422e82220a937f7d60c", "filename"=>"b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "mime_type"=>"image/jpeg"}}>} isn't valid uploaded file data
from /Users/aried3r/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/shrine-2.8.0/lib/shrine.rb:733:in `initialize'
```

```rb
=> #<MessageImage:0x00007fa8b26f8850> {
                 :id => 12150,
         :created_at => Tue, 03 Oct 2017 11:21:01 CEST +02:00,
         :updated_at => Mon, 13 Nov 2017 19:56:16 CET +01:00,
              :image => {
       :thumb => #<ImageUploader::UploadedFile:0x00007fa8b26e52a0 @data={"id"=>"private/uploads/message_image/image/12150/thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "storage"=>"store", "metadata"=>{"size"=>25772, "width"=>276, "height"=>368, "sha256"=>"ccfe5199e267fd45540d09dac6b85c5893978ea314a5493a2a86a36825875eb5", "filename"=>"thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "mime_type"=>"image/jpeg"}}>,
    :original => #<ImageUploader::UploadedFile:0x00007fa8b26e5138 @data={"id"=>"private/uploads/message_image/image/12150/b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "storage"=>"store", "metadata"=>{"size"=>550447, "width"=>1440, "height"=>1920, "sha256"=>"04b3c07231f78134998b624a3f6e00d64fb9b1755f60f422e82220a937f7d60c", "filename"=>"b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg", "mime_type"=>"image/jpeg"}}>
  },
         :message_id => 11236,
  :client_created_at => Tue, 03 Oct 2017 11:20:43 CEST +02:00,
         :image_data => {
       "thumb" => {
            "id" => "private/uploads/message_image/image/12150/thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg",
       "storage" => "store",
      "metadata" => {
             "size" => 25772,
            "width" => 276,
           "height" => 368,
           "sha256" => "ccfe5199e267fd45540d09dac6b85c5893978ea314a5493a2a86a36825875eb5",
         "filename" => "thumb_b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg",
        "mime_type" => "image/jpeg"
      }
    },
    "original" => {
            "id" => "private/uploads/message_image/image/12150/b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg",
       "storage" => "store",
      "metadata" => {
             "size" => 550447,
            "width" => 1440,
           "height" => 1920,
           "sha256" => "04b3c07231f78134998b624a3f6e00d64fb9b1755f60f422e82220a937f7d60c",
         "filename" => "b0f7b5d6-3884-40aa-9561-2f635fd778ba.jpg",
        "mime_type" => "image/jpeg"
      }
    }
  }
}
```

Reply:
`Shrine::UploadedFile#refresh_metadata!` will just refresh the metadata of the `UploadedFile` object in memory, it will not save the record with the updated `UploadedFile` object. This should do the trick:

```rb
Photo.find_each do |photo|
  attachment = Shrine.uploaded_file(photo.image, &:refresh_metadata!)
  photo.update(image_data: attachment.to_json)
end
```

It would probably be useful to add a `Shrine::Attacher#refresh_metadata!` method which writes the uploaded files to the column attribute (but doesn't save the record), and then one could do:

```rb
Photo.find_each do |photo|
  photo.image_attacher.refresh_metadata!
  photo.save
end
```

Reply:
Thanks, I'll try that!

I am still wondering why I couldn't get `Shrine.uploaded_file(message_image.image)` to work. Any clues? I tried passing in the UploadedFile as well as the `image_data` column to no avail.

Also the proposed `Shrine.uploaded_file(photo.image, &:refresh_metadata!)` doesn't work for the same reason.

Reply:
@aried3r I only now actually read the exception your received. It's because you loaded the `versions` plugin only for the `ImageUploader`, not `Shrine`. `Shrine.uploaded_file` method will know how to loop through a hash of `Shrine::UploadedFile` objects only if the `versions` plugin is loaded, which overrides that method.

I see how this can be surprising behaviour, I don't have an idea yet how to improve this user experience.

Reply:
Aaaaah, I was always looking at [`shrine.rb`](https://github.com/janko-m/shrine/blob/247f75ffd2cc27051c74337dda2da17c1027d35c/lib/shrine.rb#L164-L180) and didn't check [`versions.rb`](https://github.com/janko-m/shrine/blob/247f75ffd2cc27051c74337dda2da17c1027d35c/lib/shrine/plugins/versions.rb#L154-L164). I was puzzled by how that was supposed to work.

Thanks for clarifying! Maybe I'll add that to the docs as well.